### PR TITLE
Liam/commit info

### PR DESCRIPTION
**Pull request by liam-ilan**
### Summary
Every board exposes to the CAN bus the current running commit hash, and if the commit is clean.

### Changelist 
- Added python script utility to generate header files that expose the short git commit hash and if the commit is clean (`scripts/code_generation/commit_info_gen`).
- Modified `firmware/CMakeLists.txt` to generate header files with `commit_info_gen`. 
- Added `BOARDNAME_CommitInfo` message to every `BOARDNAME_tx.json`.
    - 32 bit `Hash`, 1 bit `Clean`.
- Every board broadcasts commit info through CAN bus.

### Testing Done
- Flashed on Car

### Resolved Issues
None

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: firmware/thruna/PDM/Src/main.c
Comment (Lucien950) [inline firmware/thruna/PDM/Src/main.c:59]: PDM does not broadcast commit info, despite importing + creating message

File: firmware/CMakeLists.txt
Comment (Lucien950) [inline firmware/CMakeLists.txt:497]: remove/complete

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (Lucien950) [inline scripts/code_generation/commit_info_gen/src/generate_commit_info.py:14]: instead of using comments, use 3 variable names.

**Review by Lucien950 — CHANGES_REQUESTED** [review]
Works on my machine. Small things before we approve. Also, `DIM` does not broadcast commit info (I didn't know where to comment that, so I'm putting it here).

File: scripts/code_generation/commit_info_gen/.gitignore
Comment (liam-ilan) [reply scripts/code_generation/commit_info_gen/.gitignore:2]: Nope, good catch! I think it's also better to include an example header in the `commit_info` directory for reference, will fix.

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (liam-ilan) [reply scripts/code_generation/commit_info_gen/src/generate_commit_info.py]: Fixed!

File: scripts/code_generation/commit_info_gen/.gitignore
Comment (liam-ilan) [reply scripts/code_generation/commit_info_gen/.gitignore]: Fixed!

File: firmware/CMakeLists.txt
Comment (liam-ilan) [reply firmware/CMakeLists.txt]: Fixed!

File: firmware/thruna/PDM/Src/main.c
Comment (liam-ilan) [reply firmware/thruna/PDM/Src/main.c:59]: Fixed! Good catch.

**Comment by Lucien950** [comment]
formatting probably murks the head file because it is not used, mentioned in review

**Comment by liam-ilan** [comment]
@Lucien950 
> Works on my machine. Small things before we approve. Also, `DIM` does not broadcast commit info (I didn't know where to comment that, so I'm putting it here).

I'll try to get CMake working to add the header for every board, rather than just the DIM, though I have a bit of an issue here...

Currently, as a quick fix, we just add the `App_CommitInfo.h` header file to `ARM_BINARY_SRCS`. Since this is a header file, it should actually be covered under `ARM_BINARY_INCLUDE_DIRS`. `ARM_BINARY_INCLUDE_DIRS` covers `Inc/App`, where `App_CommitInfo.h` is, however the commit info header file is not picked up... I'm not sure why...

Adding `App_CommitInfo.h` as a source, and not an include, works for all boards pre-refactor (everything but the `DIM`).

Does every autogenerated header file need an accompanying source file to be scanned as a dependency? In this case said source file would just include the header, and be empty... But if that's the fix, I can do it.

I'll try to experiment with this.

**Comment by liam-ilan** [comment]
> formatting probably murks the head file because it is not used, mentioned in review

CI checks currently fail on the PDM on the `git diff`... Ran clang formatting... Unsure what the issue is...

**Comment by Lucien950** [comment]
> > formatting probably murks the head file because it is not used, mentioned in review
> 
> CI checks currently fail on the PDM on the `git diff`... Ran clang formatting... Unsure what the issue is...

You have put the functions before the zone alloted for user code. Put it after this line
https://github.com/UBCFormulaElectric/Consolidated-Firmware/blob/a55c6cc64360a5b3048f37fee08c65c1a436e2c9/firmware/thruna/PDM/Src/main.c#L193

**Comment by liam-ilan** [comment]
Everything should work on all boards, no exceptions! Next step is to test the software on the car itself...

File: firmware/CMakeLists.txt
Comment (Lucien950) [inline firmware/CMakeLists.txt:28]: perhaps set this as a global variable, so it doesn't get regenerated on each run, less side effects

Comment (Lucien950) [reply firmware/CMakeLists.txt:28]: or possibly just inline it on L33 (but that would make it slightly less readable)

File: firmware/thruna/BMS/Src/main.c
Comment (Lucien950) [inline firmware/thruna/BMS/Src/main.c:282]: remove extra comments?

**Review by Lucien950 — CHANGES_REQUESTED** [review]
other than a few small things to clean up, it is done

**Review by Lucien950 — APPROVED** [review]
once this is tested, this is solid

**Comment by liam-ilan** [comment]
Current status:
- All boards but FSM work.
  - FSM Watchdog Times Out, need to fix. **DO NOT MERGE YET**
- PCAN displays the hash in decimal. Viewing raw hexadecimal results in a sort of reversed hash (eg. `9abcdef` becomes `ef cd ab 09`). I confirmed that the correct hash is being transmitted, this is a formatting issue.
  - Currently unsure if we can change the radix via the dbc file... If we can't, we need to stop PCAN from reversing raw hex data. If this is also not possible, the system is still super helpful, it's just a bit more clunky to read.

**Comment by gtaharaedmonds** [comment]
> > > formatting probably murks the head file because it is not used, mentioned in review
> > 
> > 
> > CI checks currently fail on the PDM on the `git diff`... Ran clang formatting... Unsure what the issue is...
> 
> You have put the functions before the zone alloted for user code. Put it after this line (line 193)
> 
> https://github.com/UBCFormulaElectric/Consolidated-Firmware/blob/a55c6cc64360a5b3048f37fee08c65c1a436e2c9/firmware/thruna/PDM/Src/main.c#L190-L198
> 
> I would also suggest adding this function (in all boards) in the 1Hz task, in order to keep refreshing it, in case the CAN viewer is not plugged in when the board starts up. @gtaharaedmonds I believe that is the case, please fact check me on that one

No when you call a CAN setter it sets a static variable (the "CAN table"). This means if you just set the value once it will persist for the lifetime of the program. For that reason, I think you should move this back to `main.c`.

**Comment by gtaharaedmonds** [comment]
> * FSM Watchdog Times Out, need to fix. **DO NOT MERGE YET**

FSM watchdog timeout has been fixed in https://github.com/UBCFormulaElectric/Consolidated-Firmware/pull/1038, I don't think is your change.

Okay that's too bad but this is still great to have. Even if its a bit more work to read its great that all of our data logs are now stamped with the commit. We could look back into doing this with strings in the future, but I don't think thats high priority.

File: scripts/code_generation/commit_info_gen/README.md
Comment (gtaharaedmonds) [inline scripts/code_generation/commit_info_gen/README.md:1]: Actual documentation is great

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (gtaharaedmonds) [inline scripts/code_generation/commit_info_gen/src/generate_commit_info.py:44]: Nit: It'd probably be better to do this as a bool and import `<stdbool.h>`

Up to you, this is very minor

**Review by gtaharaedmonds — APPROVED** [review]
LGTM. Can you move the CAN setters back to `main.c`, and then let's merge!

File: scripts/code_generation/commit_info_gen/README.md
Comment (Lucien950) [reply scripts/code_generation/commit_info_gen/README.md:1]: also on confluence?

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (Lucien950) [reply scripts/code_generation/commit_info_gen/src/generate_commit_info.py:44]: I would prefer booleans in this case

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (Lucien950) [inline scripts/code_generation/commit_info_gen/src/generate_commit_info.py:50]: i would suggest catching the specific error here we are expecting in order to allow other errors to stop the process

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (liam-ilan) [reply scripts/code_generation/commit_info_gen/src/generate_commit_info.py]: Now using booleans!

File: scripts/code_generation/commit_info_gen/README.md
Comment (gtaharaedmonds) [reply scripts/code_generation/commit_info_gen/README.md:1]: imo its fine to stay in the repo since this is implementation details, I don't think another team will have to look at this file

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (gtaharaedmonds) [inline scripts/code_generation/commit_info_gen/src/generate_commit_info.py:20]: nit: Python convention is to use `snake_case` for variables and function names, and `PascalCase` for classes

File: can_bus/json/BMS/BMS_tx.json
Comment (gtaharaedmonds) [inline can_bus/json/BMS/BMS_tx.json:327]: You can review the board name prefixes, they're added automatically now. 
This change just went in today so I still need to make a post about it

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (liam-ilan) [reply scripts/code_generation/commit_info_gen/src/generate_commit_info.py:53]: Tried to make things a bit clearer in commit e55adf4249ba968141137c430e981a249d7c51a1.

Previously, one try catch would handle both GitPython and data validation. Now data validation is not in the try except, and the try except handles GitPython errors only.

Because commit_info_gen is non-critical, we want to stop it from blocking compilation as much as possible. To do this, if any error is released by GitPython, dummy data is returned instead, with a warning logged to the terminal. We don't care what type of error is thrown, all errors should be caught so the process can continue.

File: scripts/code_generation/commit_info_gen/src/generate_commit_info.py
Comment (liam-ilan) [reply scripts/code_generation/commit_info_gen/src/generate_commit_info.py:20]: good catch, fixed!

File: can_bus/json/BMS/BMS_tx.json
Comment (liam-ilan) [reply can_bus/json/BMS/BMS_tx.json]: Fixed!

**Review by Lucien950 — APPROVED** [review]

**Pull request merged**
